### PR TITLE
fix: input file encoding

### LINKED ISSUE #448: Error getting BOM file from poetry
I have a project in which I use poetry, and I want to be able to obtain the BOM file with the poetry dependencies.

When I run the following command: `cyclonedx-py --poetry --output bom.xml`

I get the following error:
```text
Traceback (most recent call last):
  File "C:\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\jcmencia\PycharmProjects\QA-Pruebas\py_requests\venv\Scripts\cyclonedx-py.exe\__main__.py", line 7, in <module>     
  File "C:\Users\jcmencia\PycharmProjects\QA-Pruebas\py_requests\venv\lib\site-packages\cyclonedx_py\client.py", line 307, in main   
    CycloneDxCmd(args).execute()
  File "C:\Users\jcmencia\PycharmProjects\QA-Pruebas\py_requests\venv\lib\site-packages\cyclonedx_py\client.py", line 147, in execute
    output = self.get_output()
  File "C:\Users\jcmencia\PycharmProjects\QA-Pruebas\py_requests\venv\lib\site-packages\cyclonedx_py\client.py", line 84, in get_output
    parser = self._get_input_parser()
  File "C:\Users\jcmencia\PycharmProjects\QA-Pruebas\py_requests\venv\lib\site-packages\cyclonedx_py\client.py", line 282, in _get_input_parser
    input_data = input_data_fh.read()
  File "C:\Python310\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 6935: character maps to <undefined>
```

The poetry.lock file is in UTF-8 format and with LF line endings and I use Windows with PyCharm

If I edit in the client.py file the following line:
```py
self._arguments.input_source = open(os.path.join(current_directory, 'poetry.lock'), 'r')
```
And I add the _encoding_ parameter:
```py
self._arguments.input_source = open(os.path.join(current_directory, 'poetry.lock'), 'r', encoding="utf-8")
````
The command works perfectly for me.

Is there any configuration I can do to make it use UTF-8 encoding instead of opting for cp1252? Or any solution for this problem? Everything I have tried has not worked.
Thanks in advance!

### PR DESCRIPTION
**Pull request by JCHacking**
input files in lock-format are expected in a certain encoding,
other input file encodings  are detected.

fixes #448

### CODE REVIEW

**Comment by jkowalleck** [comment]
why did you use a niche library https://pypi.org/project/faust-cchardet
why not use a library that is widely used? like https://pypi.org/project/chardet ?

**Comment by JCHacking** [comment]
> why did you use a niche library https://pypi.org/project/faust-cchardet
> why not use a library that is widely used? like https://pypi.org/project/chardet ?
> 
> if it was for the py3.6 compatibility, `chardet` provides older compatible versions.

That was one of the reasons, the other one is because some encodings like cp1252 are returned as Windows-1252 so they are not exactly the same to pass it to the decode method.

Also faust-cchardet is a maintained fork of cchardet which is written in C so it has better performance.
Traducido con DeepL https://www.deepl.com/app/?utm_source=android&utm_medium=app&utm_campaign=share-translation

**Comment by jkowalleck** [comment]
re https://github.com/CycloneDX/cyclonedx-python/pull/596#issuecomment-1770403331
> [...] it has better performance

Performance should not be a concern at this point.
Maintainability and support of used libraries is more important here. So better use a well-maintained library like `chardet` that is widely used. 
If this choice is insufficient at any point, we can still report bugs to the maintainers team of that library. 

----

see https://github.com/CycloneDX/cyclonedx-python/issues/448#issuecomment-1770454853

could you test the following on your system? 
- for `poetry.lock` and `Pipfile.lock` use: `open(..., encoding="utf8")` -- no mode, must path and encoding
- for `requirements.txt` the char-detection is okay in general.

**Comment by JCHacking** [comment]
> re [#596 (comentario)](https://github.com/CycloneDX/cyclonedx-python/pull/596#issuecomment-1770403331)
> 
> > [...] tiene mejor rendimiento
> 
> El rendimiento no debería ser una preocupación en este momento. La mantenibilidad y el soporte de las bibliotecas usadas son más importantes aquí. Así que es mejor utilizar una biblioteca bien mantenida como `chardet`la que se utiliza ampliamente. Si esta elección es insuficiente en algún momento, aún podemos informar errores al equipo de mantenimiento de esa biblioteca.
> 
> ver [#448 (comentario)](https://github.com/CycloneDX/cyclonedx-python/issues/448#issuecomment-1770454853)
> 
> ¿Podrías probar lo siguiente en tu sistema?
> 
> * para `poetry.lock`y `Pipfile.lock`uso: `open(..., encoding="utf8")`-- sin modo, debe ruta y codificación
> * para `requirements.txt`la detección de caracteres está bien en general.

Perfect, then I change the library to chardet with a version that supports python 3.6 and then I try the other option that you mention of only validating the coding in requirements.txt

**Comment by JCHacking** [comment]
The change is already done

Basically, if the open is done in byte mode the encoding will be inspected, otherwise it will be assumed that everything is OK.

And regarding chardet I have made a replace to make it work with windows, since it returns Windows-1252 but python only understands cp1252.

File: pyproject.toml [hunk 1]
```diff
@@ -50,6 +50,10 @@ importlib-metadata = { version = ">= 3.4", python = "< 3.8" }
 pip-requirements-parser = "^32.0.0"
 setuptools = ">= 47.0.0"
 toml = "^0.10.0"
+chardet = [
+  {version = "^4.0", python = "=3.6"},
```
Comment (jkowalleck) [inline pyproject.toml:54]: {version = "^4.0", python = "~3.6.0"},
{version = "^5.2", python = "^3.7"},

**Comment by jkowalleck** [comment]
I had to do minor version range adjustments and other chores.

I will add a regression test, and then, this fix is ready to go.
Thank you for your effort, @JCHacking

**Review by jkowalleck — APPROVED** [review]

**Comment by JCHacking** [comment]
> I had to do minor version range adjustments and other chores.
> 
> I will add a regression test, and then, this fix is ready to go.
> Thank you for your effort, @JCHacking 

Thanks to you for letting me collaborate in its solution.

**Pull request merged**